File: services/temporal_gateway/internal/gateway/service.go
package gateway

import (
	"context"
	"fmt"

	workflowv1 "github.com/nshkrdotcom/alkahest/services/temporal_gateway/gen/alkahest/workflow/v1"
	enums "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/client"
)

// Service implements the Alkahest gRPC gateway contract over the official Temporal Go SDK.
type Service struct {
	workflowv1.UnimplementedWorkflowRuntimeGatewayServer
	client    client.Client
	namespace string
}

// NewService builds a gateway service backed by a Temporal client.
func NewService(temporalClient client.Client, namespace string) *Service {
	return &Service{client: temporalClient, namespace: namespace}
}

// StartWorkflow starts a workflow execution through Temporal.
func (s *Service) StartWorkflow(ctx context.Context, req *workflowv1.StartWorkflowRequest) (*workflowv1.StartWorkflowResponse, error) {
	workflowRun, err := s.client.ExecuteWorkflow(
		ctx,
		client.StartWorkflowOptions{
			ID:                    req.GetWorkflowId(),
			TaskQueue:             req.GetTaskQueue(),
			WorkflowIDReusePolicy: enums.WORKFLOW_ID_REUSE_POLICY_REJECT_DUPLICATE,
		},
		req.GetWorkflowType(),
		req.GetInputJson(),
	)
	if err != nil {
		return &workflowv1.StartWorkflowResponse{Error: normalizeError(err)}, nil
	}

	return &workflowv1.StartWorkflowResponse{
		Ref: &workflowv1.WorkflowRef{
			Namespace:  s.namespace,
			WorkflowId: workflowRun.GetID(),
			RunId:      workflowRun.GetRunID(),
		},
		Status: "started",
	}, nil
}

// SignalWorkflow sends a signal to a workflow execution.
func (s *Service) SignalWorkflow(ctx context.Context, req *workflowv1.SignalWorkflowRequest) (*workflowv1.SignalWorkflowResponse, error) {
	ref := req.GetRef()
	if err := s.client.SignalWorkflow(ctx, ref.GetWorkflowId(), ref.GetRunId(), req.GetSignalName(), req.GetPayloadJson()); err != nil {
		return &workflowv1.SignalWorkflowResponse{Ref: ref, Error: normalizeError(err)}, nil
	}

	return &workflowv1.SignalWorkflowResponse{Ref: ref, Status: "signaled"}, nil
}

// QueryWorkflow queries a workflow execution.
func (s *Service) QueryWorkflow(ctx context.Context, req *workflowv1.QueryWorkflowRequest) (*workflowv1.QueryWorkflowResponse, error) {
	ref := req.GetRef()
	value, err := s.client.QueryWorkflow(ctx, ref.GetWorkflowId(), ref.GetRunId(), req.GetQueryName(), req.GetPayloadJson())
	if err != nil {
		return &workflowv1.QueryWorkflowResponse{Ref: ref, Error: normalizeError(err)}, nil
	}

	var result string
	if err := value.Get(&result); err != nil {
		return &workflowv1.QueryWorkflowResponse{Ref: ref, Error: normalizeError(err)}, nil
	}

	return &workflowv1.QueryWorkflowResponse{Ref: ref, ResultJson: result}, nil
}

// CancelWorkflow requests workflow cancellation.
func (s *Service) CancelWorkflow(ctx context.Context, req *workflowv1.CancelWorkflowRequest) (*workflowv1.CancelWorkflowResponse, error) {
	ref := req.GetRef()
	if err := s.client.CancelWorkflow(ctx, ref.GetWorkflowId(), ref.GetRunId()); err != nil {
		return &workflowv1.CancelWorkflowResponse{Ref: ref, Error: normalizeError(err)}, nil
	}

	return &workflowv1.CancelWorkflowResponse{Ref: ref, Status: "cancel_requested"}, nil
}

// DescribeWorkflow returns a compact workflow description.
func (s *Service) DescribeWorkflow(ctx context.Context, req *workflowv1.DescribeWorkflowRequest) (*workflowv1.DescribeWorkflowResponse, error) {
	ref := req.GetRef()
	description, err := s.client.DescribeWorkflowExecution(ctx, ref.GetWorkflowId(), ref.GetRunId())
	if err != nil {
		return &workflowv1.DescribeWorkflowResponse{Ref: ref, Error: normalizeError(err)}, nil
	}

	status := "unknown"
	if description.GetWorkflowExecutionInfo() != nil {
		status = description.GetWorkflowExecutionInfo().GetStatus().String()
	}

	return &workflowv1.DescribeWorkflowResponse{Ref: ref, Status: status}, nil
}

// FetchWorkflowHistoryRef returns a compact history URI rather than raw history events.
func (s *Service) FetchWorkflowHistoryRef(_ context.Context, req *workflowv1.FetchWorkflowHistoryRefRequest) (*workflowv1.FetchWorkflowHistoryRefResponse, error) {
	ref := req.GetRef()
	return &workflowv1.FetchWorkflowHistoryRefResponse{
		Ref:        ref,
		HistoryUri: fmt.Sprintf("temporal://%s/%s/%s", ref.GetNamespace(), ref.GetWorkflowId(), ref.GetRunId()),
	}, nil
}

func normalizeError(err error) *workflowv1.WorkflowError {
	return &workflowv1.WorkflowError{Code: "temporal_error", Message: err.Error(), Retryable: true}
}

File: apps/alkahest_contracts/lib/alkahest/proto/workflow/v1/service.ex
defmodule Alkahest.Proto.Workflow.V1.WorkflowRuntimeGateway.Service do
  @moduledoc "gRPC service definition for the Alkahest workflow runtime gateway."
  use GRPC.Service,
    name: "alkahest.workflow.v1.WorkflowRuntimeGateway",
    protoc_gen_elixir_version: "0.16.0"

  alias Alkahest.Proto.Workflow.V1.CancelWorkflowRequest
  alias Alkahest.Proto.Workflow.V1.CancelWorkflowResponse
  alias Alkahest.Proto.Workflow.V1.DescribeWorkflowRequest
  alias Alkahest.Proto.Workflow.V1.DescribeWorkflowResponse
  alias Alkahest.Proto.Workflow.V1.FetchWorkflowHistoryRefRequest
  alias Alkahest.Proto.Workflow.V1.FetchWorkflowHistoryRefResponse
  alias Alkahest.Proto.Workflow.V1.QueryWorkflowRequest
  alias Alkahest.Proto.Workflow.V1.QueryWorkflowResponse
  alias Alkahest.Proto.Workflow.V1.SignalWorkflowRequest
  alias Alkahest.Proto.Workflow.V1.SignalWorkflowResponse
  alias Alkahest.Proto.Workflow.V1.StartWorkflowRequest
  alias Alkahest.Proto.Workflow.V1.StartWorkflowResponse

  rpc(:StartWorkflow, StartWorkflowRequest, StartWorkflowResponse)
  rpc(:SignalWorkflow, SignalWorkflowRequest, SignalWorkflowResponse)
  rpc(:QueryWorkflow, QueryWorkflowRequest, QueryWorkflowResponse)
  rpc(:CancelWorkflow, CancelWorkflowRequest, CancelWorkflowResponse)
  rpc(:DescribeWorkflow, DescribeWorkflowRequest, DescribeWorkflowResponse)
  rpc(:FetchWorkflowHistoryRef, FetchWorkflowHistoryRefRequest, FetchWorkflowHistoryRefResponse)
end

defmodule Alkahest.Proto.Workflow.V1.WorkflowRuntimeGateway.Stub do
  @moduledoc "gRPC client stub for the Alkahest workflow runtime gateway."
  use GRPC.Stub, service: Alkahest.Proto.Workflow.V1.WorkflowRuntimeGateway.Service
end
